### PR TITLE
Exclude visual elements

### PR DESCRIPTION
**Pull request by dbcesar**

### CODE REVIEW

File: src/Vizkit3dWorld.cpp [hunk 1]
```diff
@@ -18,8 +18,12 @@
 
 namespace vizkit3d_world {
 
-Vizkit3dWorld::Vizkit3dWorld(std::string path, std::vector<std::string> modelPaths,
-                            int cameraWidth, int cameraHeight, double horizontalFov, double zNear, double zFar)
+Vizkit3dWorld::Vizkit3dWorld(std::string path, 
+                            std::vector<std::string> modelPaths,
+                            std::vector<std::string> exclude_from_sdf,
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: One golden rule about naming convention: stick to whatever the package already chose. Here, this means that you should really use lowerCamelCase for parameter names.

File: src/Vizkit3dWorld.cpp [hunk 2]
```diff
@@ -270,11 +287,9 @@ void Vizkit3dWorld::setCameraPose(base::samples::RigidBodyState pose) {
      * - Y left
      * - Z up
      */
-    osg::Matrixd m = poseMatrix *
-                     osg::Matrixd::rotate(M_PI_2, osg::Vec3(0.0, 0.0, 1.0)) *
-                     osg::Matrixd::rotate(-M_PI_2, osg::Vec3(1.0, 0.0, 0.0));
-
-
+    osg::Matrixd m = poseMatrix*
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp:280]: What does this have to do with the model exclusion ?

File: src/Vizkit3dWorld.cpp (hunk 1 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Done!

File: src/Vizkit3dWorld.cpp (hunk 2 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp:280]: Again here...nothing related to the model exclusion, but with the process of SDF updating.
Both visual exclusion and that transformation is related to handle with the changes done in the sdf model. Is it recommended to open a different PR to deal with it?

Comment (doudou) [inline src/Vizkit3dWorld.cpp:280]: You need to at least put it in a different commit that explains the change. Given how small the changes are, a different PR is not needed.

File: src/Vizkit3dWorld.cpp [hunk 3]
```diff
@@ -330,4 +345,38 @@ void Vizkit3dWorld::applyCameraParams() {
     widget->getView(0)->getCamera()->setProjectionMatrixAsPerspective(osg::RadiansToDegrees(fovy), aspectRatio, zNear, zFar);
 }
 
+void Vizkit3dWorld::removeLinksChild(sdf::ElementPtr sdf_model, std::string child_element){
+
+    if (sdf_model->HasElement("link"))
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: You don't need the two-stage if / while here. if the model does not have the element, linkElem will be NULL which is then ignored by while(). Just remove the if

Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Done

File: src/Vizkit3dWorld.cpp (hunk 2 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp:280]: Done!

File: src/Vizkit3dWorld.hpp [hunk 4]
```diff
@@ -133,6 +134,12 @@ class Vizkit3dWorld : public QObject {
     void loadGazeboModelPaths(std::vector<std::string> modelPaths = std::vector<std::string>());
 
     /**
+     * Load the list of the models that we do not want to have the visual nor collision elements
+     * @params list of model's name
+     */
+    void loadExlcudedSdfModels(std::vector<std::string> excludeFromSdf = std::vector<std::string>());
```
Comment (doudou) [inline src/Vizkit3dWorld.hpp]: Typo in method name ("Excluded"). Moreover, there's really no point in having a default argument here (why would one want to call `loadedExcludedModels()` ?)

File: src/Vizkit3dWorld.cpp [hunk 5]
```diff
@@ -155,6 +164,14 @@ vizkit3d::RobotVisualization* Vizkit3dWorld::robotVizFromSdfModel(sdf::ElementPt
     //vizkit3d plugin with model defined in the sdf model
     vizkit3d::RobotVisualization* robotViz = new vizkit3d::RobotVisualization();
 
+    for (int i=0; i < excludeFromSdf.size(); ++i){
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: One thing I just realized: why do you exclude the visual/collision links instead of ignoring the model ?

Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: To be honest, since ignoring the whole model will ignore the transformation links I don't know how it will impact the simulation. I've talked with Gustavo and he also thinks that it will affect, but he is not sure as well. So if you see a relevant difference between ignore the model and remove the visual/collisions I can try and verify what will happen.

Comment (doudou) [inline src/Vizkit3dWorld.cpp]: I was asking because I don't see any. Except that ignoring the whole model is obviously a LOT simpler. So, I would go for the simple solution and go back to that if the simple solution has issues.

Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: @doudou , I've tried to ignore the entire model, but it didn't work. I've got the following message.

error

Since the links corresponds to a frame, it seems that I have to load at least them. Because I think Vizkit3d fails when it tries to find the frame names.

Comment (doudou) [inline src/Vizkit3dWorld.cpp]: It's weird that the camera frame changes to flat_fish. There's no reason for that since we definitely do NOT place the camera there. Would be good to know why this error pops up.

File: src/Vizkit3dWorld.cpp [hunk 6]
```diff
@@ -330,4 +345,34 @@ void Vizkit3dWorld::applyCameraParams() {
...
+                undesired_elem = undesired_elem->GetNextElement(child_element);
+
+            }
+
+            for(int i=0; i < erase_vector.size(); ++i)
+            {
+                linkElem->RemoveChild(erase_vector[i]);
+            }
+
+            erase_vector.clear();
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: Why do you clear the vector ?

File: src/Vizkit3dWorld.cpp [hunk 7]
```diff
@@ -330,4 +345,34 @@ void Vizkit3dWorld::applyCameraParams() {
     widget->getView(0)->getCamera()->setProjectionMatrixAsPerspective(osg::RadiansToDegrees(fovy), aspectRatio, zNear, zFar);
 }
 
+void Vizkit3dWorld::removeLinksChild(sdf::ElementPtr sdf_model, std::string child_element){
+
+    sdf::ElementPtr linkElem = sdf_model->GetElement("link");
+
+    while(linkElem){
+        if (linkElem->HasElement(child_element)){
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: The if/while here is redundant. Nothing will happen if GetElement(child_element) is NULL, so just remove the if

File: src/Vizkit3dWorld.cpp [hunk 8]
```diff
@@ -330,4 +345,34 @@ void Vizkit3dWorld::applyCameraParams() {
...
+            sdf::ElementPtr undesired_elem = linkElem->GetElement(child_element);
+
+            //stores all the undesired elements before delete them
+            std::vector<sdf::ElementPtr> erase_vector;
+            while (undesired_elem)
+            {
+                erase_vector.push_back(undesired_elem);
+
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: Don't add empty lines without good reasons. Think of empty lines as separating paragraphs. The same way than you would split paragraphs when writing text, you need to try and keep code that logically belongs together, together.

File: src/Vizkit3dWorld.cpp [hunk 9]
```diff
@@ -330,4 +345,34 @@ void Vizkit3dWorld::applyCameraParams() {
...
+                undesired_elem = undesired_elem->GetNextElement(child_element);
+
```
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: And definitely do NOT add empty spaces before a block end / after a block beginning.

File: src/Vizkit3dWorld.cpp (hunk 5 above)
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: Ping. Understanding this _is_ important.

File: src/Vizkit3dWorld.cpp (hunk 7 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Done

File: src/Vizkit3dWorld.cpp (hunk 5 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Checking it now

File: src/Vizkit3dWorld.cpp (hunk 6 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Currently there is no reason. Before I removed the visual and collision in the same while block,  but I changed it later and forgot to delete it.

File: src/Vizkit3dWorld.cpp (hunk 9 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Done

File: src/Vizkit3dWorld.cpp (hunk 8 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Got it! Thx for the advice!

File: src/Vizkit3dWorld.cpp (hunk 7 above)
Comment (doudou) [inline src/Vizkit3dWorld.cpp]: > Done

Still there ...

Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Sorry... now it is done.

File: src/Vizkit3dWorld.cpp (hunk 5 above)
Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Hi....I still didn't figure out why the pop up appears. I'll go deeper on that tomorrow, but it happen not specifically with flat_fish, but with the sdf that we want to exclude (see the mega_structure case)
erro2

However after I click the "ok" button on the pop up the camera starts to run normally. It ignores the sdf files now (I realized that was doing some stuffs wrong yesterday).

So, I think your suggestion of ignoring the sdf instead of removing the visual and collision elements is the best way, but I still need to figure out how to prevent the pop up messages.

Comment (doudou) [inline src/Vizkit3dWorld.cpp]: Hunch: the component is still provided with the pose of the model you exclude, and you get the popup when the RigidBodyState is processed.

Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: Ok! I think I fixed it.

I've noticed that every time the property "frame" is changed in the RobotVisualization, a callback function checks if this visualizationFrame is part of any segment of the model. (not sure here, but I think that it defines the frame we see when we change the ViewerFrame on the vizkit widget)
So what happens is I was setting a visualisationFrame that wasn't even a valid segment, since I didn't load the sdfmodel to the RobotViz, and that's why the popup was called.
What I did was to set the "world_osg" for the RobotViz that we want to ignores. I'm pushing the code in a few minutes.

Comment (dbcesar) [inline src/Vizkit3dWorld.cpp]: @doudou, I've already pushed it  I don't know the whole Vizkit integration mechanism so it is important to see if what I did can impact somehow the other functionalities, for instance, ignoring important information for the other parts of the code. Locally, as far as I could see everything is running ok.

**Comment by doudou** [comment]
If  you don't visualize the model, why do you even consider it ? Intuitively, the filtering should happen before the call to `robotVizFromSdfModel` (which removes the need for a special case in `attachPlugins`)

**Comment by doudou** [comment]
Ping

**Comment by dbcesar** [comment]
Hi, 
I had tried it before and just tested again now to filter the model before  `robotVizFromSdfModel`. What happens is that the model is not inserted inside the `toSdfElement` std::map and the system crashes when I call the camera task. The error message is:

`Syskit::PortNotFound: OroGen::FlatFishCameraSimulation::FlatfishCamera:0x91350e8(conf: ["default"], device_dev: device(Dive::Devices::Simulation::UnderwaterCamera, as: raw_front_left_camera), orocos_name: front_left_camera)[] has connections but there are missing ports on the actual RTT task`

Inside the componet, I get the following message:

```
0.088 [ Warning][TaskContextServer()] CTaskContext 'front_left_camera' already bound to CORBA Naming Service.
0.088 [ Warning][TaskContextServer()] Trying to rebind... done. New CTaskContext bound to Naming Service.
0.089 [ Warning][TaskContextServer()] CTaskContext 'front_left_camera_Logger' already bound to CORBA Naming Service.
0.089 [ Warning][TaskContextServer()] Trying to rebind... done. New CTaskContext bound to Naming Service.
WARNING: QApplication was not created in the main() thread.
Inconsistency detected by ld.so: dl-close.c: 762: _dl_close: Assertion `map->l_init_called' failed!
```

**Comment by doudou** [comment]
Two problems here

The technical one: Syskit does not know about the excluded models. It assumes that you will have all models set up, and yells because the flat_fish model is not. Could you please update the PR so that I or @gustavoneves12 adapt the Syskit part ?

The conceptual one: you thought about doing the right thing. If it does not work right away, do NOT drop the idea "just because another idea will work". Find out _why_ your "right" idea did not work (possibly asking around in this case, since it involves syskit).

If you always take the path to the quick workaround, you will never really understand how the system you're working on _works_. You'll just have a pile of magical workarounds, and at the first complex problem, you'll be dead in the water.

**Comment by dbcesar** [comment]
> Could you please update the PR so that I or @gustavoneves12 adapt the Syskit part ?

Done!

> The conceptual one: you thought about doing the right thing. If it does not work right away, do NOT drop the idea "just because another idea will work". Find out why your "right" idea did not work (possibly asking around in this case, since it involves syskit).

Understood!

**Comment by dbcesar** [comment]
Ping. Just to know it is which one of you @doudou or @gustavoneves12 are going to work on that.

**Comment by doudou** [comment]
I'm finishing up some test-related stuff on the bundles, I'll have a look afterwards.
